torchtune/utils/_activations_offloading/offload_ctx_mgr.py

```diff
@@ -0,0 +1,260 @@
+        self.is_first_forward_call = True
+        self.is_first_backward_call = True
+        self.is_first_forward_pass = True
+
+        # managing cpu memory
+        self.use_pin_memory: bool = use_pin_memory
+        self.virtual_memory_safe_pct = (
+            0.60  # we should not exceed this percentage of memory
+        )
+
+        # for streaming
+        self.s0 = torch.cuda.default_stream()  # comp stream
+        self.s1 = torch.cuda.Stream()  # comms stream
+        self.fwd_stash = {}  # tensor_id => (activation, ev1)
+        if max_fwd_stash_size < 1:
+            raise ValueError(
+                f"max_fwd_stash_size should be at least 1 but is {max_fwd_stash_size}"
+            )
+        self.max_fwd_stash_size = max_fwd_stash_size
+        self.bwd_tensor_stash = {}  # tensor_id => activation
+        self.bwd_ev_stash = {}  # tensor_id => ev0
+        self.curr_graph_id = None
+        self.curr_autograd_node = None
+
+        # platform util functions
+        def verify_sufficient_virtual_memory():
+            curr_pct = get_cpu_ram_pct()
+            if curr_pct > self.virtual_memory_safe_pct:
+                print(
+                    f"***** WARNING: {curr_pct=}% > {self.virtual_memory_safe_pct=}% of virtual memory used"
+                )
+
+        def get_cpu_ram_pct() -> float:
+            # get the percentage of memory used by the system
+            return psutil.virtual_memory().percent
+
+        def get_tensor_id() -> int:
+            # create a unique id for each tensor we are managing
+            self.tensor_id += 1
+            return self.tensor_id
+
+        def get_num_bytes_tensor(x: torch.Tensor) -> int:
+            # get the number of bytes in a tensor, for memory management purposes
+            return (
+                x.element_size() * x.nelement()
+            )  # x.element_size() * x._base_storage().nbytes()
+
+        # -------- core pack / unpack work --------
+        def pack_tensor(activation: torch.Tensor) -> int:
+            # activations are passed in during forward pass - from here we take over and return a unique id
+            if self.is_first_forward_call:
+                assert (
+                    len(self.tracker) == 0
+                ), "backward pass should have cleared tracker of all tensors"
+
+                # set training phase trackers
+                self.is_first_forward_call = False
+                self.is_first_backward_call = True
+
+            # query for basic tensor info
+            num_bytes = get_num_bytes_tensor(activation)
+            tensor_id = get_tensor_id()
+
+            # only offload hefty bois
+            if num_bytes >= self.min_tensor_size_bytes:
+                # First, sync back and dereference previously offloaded tensors
+                # as the offloading should be done sufficiently long ago.
+                for k in [x for x in self.fwd_stash.keys()]:
+                    if k <= tensor_id - self.max_fwd_stash_size:
```

Comment:
as a rule of thumb, if we can replace "x", "k" and "ev" with meaningful var names, i think it is worth to do so